daemon/input.c

```diff
@@ -0,0 +1,315 @@
+ * 
+ * @param int in    The input keyboard scan code value
+ * @param char* out Its ASCII digit representation
+ * @return          0 in case of success, -1 if the scan code value was not a digit code
+ */
+static int
+input_event_code_to_ascii_num(int in, char *out)
+{
+	if (in == KEY_1 || in == KEY_KP1) {
+		*out = '1';
+		return 0;
+	} else if (in == KEY_2 || in == KEY_KP2) {
+		*out = '2';
+		return 0;
+	} else if (in == KEY_3 || in == KEY_KP3) {
+		*out = '3';
+		return 0;
+	} else if (in == KEY_4 || in == KEY_KP4) {
+		*out = '4';
+		return 0;
+	} else if (in == KEY_5 || in == KEY_KP5) {
+		*out = '5';
+		return 0;
+	} else if (in == KEY_6 || in == KEY_KP6) {
+		*out = '6';
+		return 0;
+	} else if (in == KEY_7 || in == KEY_KP7) {
+		*out = '7';
+		return 0;
+	} else if (in == KEY_8 || in == KEY_KP8) {
+		*out = '8';
+		return 0;
+	} else if (in == KEY_9 || in == KEY_KP9) {
+		*out = '9';
+		return 0;
+	} else if (in == KEY_0 || in == KEY_KP0) {
+		*out = '0';
+		return 0;
+	} else {
+		return -1;
+	}
+}
+
+/**
+ * Parses /proc/bus/input/devices to find the /dev/input/eventx input file for the 
+ * usb devices specified by vendor_id and product_id
+ * 
+ * @param vendor_id     The vendor ID of the usb device to get the input file for
+ * @param product_id    The product ID of the usb device to get the input file for
+ * @return              The /dev/input/eventx file of the usb device
+ */
+static char *
+input_get_usb_input_event_file(uint16_t vendor_id, uint16_t product_id)
+{
+	FILE *fp;
+	char *line = NULL;
+	size_t len = 0;
+	ssize_t n;
+	uint16_t bus;
+	uint16_t vid;
+	uint16_t pid;
+	uint16_t ver;
+	char *event = NULL;
+	char *event_path = NULL;
+
+	fp = fopen("/proc/bus/input/devices", "r");
+	IF_NULL_RETVAL(fp, NULL);
+
+	while ((n = getline(&line, &len, fp)) != -1) {
+		if (strncmp(IFACE_PREFIX, line, strlen(IFACE_PREFIX)) == 0) {
+			sscanf(line, "I: Bus=%04hx Vendor=%04hx Product=%04hx Version=%04hx\n",
+			       &bus, &vid, &pid, &ver);
+		}
+
+		if ((vid == vendor_id) && (pid == product_id) &&
+		    (strncmp(HANDLER_PREFIX, line, strlen(IFACE_PREFIX)) == 0) &&
+		    strstr(line, SYSRQ_EVENT)) {
+			event = strtok(strstr(line, EVENT), " ");
+			event_path = mem_printf("/dev/input/%s", event);
+			break;
+		}
+	}
+
+	fclose(fp);
+	mem_free(line);
+	return event_path;
+}
+
+/**
+ * Read a pin from an external usb pin reader device fd points to
+ * 
+ * @param fd    The file descriptor that represents the input file
+ * @return      The read in pin
+ */
+static char *
+input_read_usb_input_event_file_pin(int fd)
+{
+	char *key = NULL;
+	char *tmp_key = NULL;
+
+	int ret = ioctl(fd, EVIOCGRAB, (void *)1);
+	if (ret != 0) {
+		WARN("Failed to get exclusive access to pin reader");
+		return key;
+	}
+	struct input_event keyboard_event;
+
+	TRACE("Reading pin from pin reader");
+	int index = 0;
+	while (1) {
+		if (read(fd, &keyboard_event, sizeof(keyboard_event)) != -1) {
+			if ((keyboard_event.type == 0x1) &&
```

Comment:
if (keyboard_event.type != 0x1 || keyboard_event.value != 0x1)
   continue;
remove "(keyboard_event.type == 0x1) && (keyboard_event.value == 0x1)" in all if cases below

Reply:
Done